File: src/tokenizer.rs
//! Host BPE tokenizer from llama3.cuda / llama2.c.
//!
//! This is generation *control* (string ↔ token ids), not a CPU transformer.

use anyhow::{bail, Context, Result};
use std::fs::File;
use std::io::Read;
use std::path::Path;

const BOS_ID: i32 = 1;
const EOS_ID: i32 = 2;

#[derive(Debug, Clone)]
pub struct Tokenizer {
    vocab: Vec<String>,
    vocab_scores: Vec<f32>,
    max_token_length: u32,
    sorted: Vec<(String, i32)>,
}

impl Tokenizer {
    pub fn from_path(path: impl AsRef<Path>, vocab_size: usize) -> Result<Self> {
        let path = path.as_ref();
        let mut file =
            File::open(path).with_context(|| format!("open tokenizer {}", path.display()))?;
        let mut max_token_length = [0u8; 4];
        file.read_exact(&mut max_token_length)
            .context("read tokenizer max_token_length")?;
        let max_token_length = u32::from_le_bytes(max_token_length);

        let mut vocab = Vec::with_capacity(vocab_size);
        let mut vocab_scores = Vec::with_capacity(vocab_size);
        for i in 0..vocab_size {
            let mut score_bytes = [0u8; 4];
            file.read_exact(&mut score_bytes)
                .with_context(|| format!("read tokenizer score {i}"))?;
            vocab_scores.push(f32::from_le_bytes(score_bytes));
            let mut len_bytes = [0u8; 4];
            file.read_exact(&mut len_bytes)
                .with_context(|| format!("read tokenizer length {i}"))?;
            let len = i32::from_le_bytes(len_bytes);
            if len < 0 {
                bail!("tokenizer piece {i} has negative length {len}");
            }
            let mut buf = vec![0u8; len as usize];
            file.read_exact(&mut buf)
                .with_context(|| format!("read tokenizer piece {i}"))?;
            vocab.push(String::from_utf8_lossy(&buf).into_owned());
        }

        Ok(Self::from_parts(vocab, vocab_scores, max_token_length))
    }

    pub fn from_parts(vocab: Vec<String>, vocab_scores: Vec<f32>, max_token_length: u32) -> Self {
        assert_eq!(vocab.len(), vocab_scores.len());
        let mut sorted: Vec<(String, i32)> = vocab
            .iter()
            .enumerate()
            .map(|(i, s)| (s.clone(), i as i32))
            .collect();
        sorted.sort_by(|a, b| a.0.cmp(&b.0));
        Self {
            vocab,
            vocab_scores,
            max_token_length,
            sorted,
        }
    }

    pub fn vocab_size(&self) -> usize {
        self.vocab.len()
    }

    pub fn max_token_length(&self) -> u32 {
        self.max_token_length
    }

    fn lookup(&self, s: &str) -> Option<i32> {
        self.sorted
            .binary_search_by(|(k, _)| k.as_str().cmp(s))
            .ok()
            .map(|i| self.sorted[i].1)
    }

    /// Encode `text`. `bos` prepends token 1; `eos` appends token 2.
    pub fn encode(&self, text: &str, bos: bool, eos: bool) -> Vec<i32> {
        let mut tokens = Vec::new();
        if bos {
            tokens.push(BOS_ID);
        }
        if !text.is_empty() {
            if let Some(dummy) = self.lookup(" ") {
                tokens.push(dummy);
            }
        }

        let bytes = text.as_bytes();
        let mut i = 0;
        while i < bytes.len() {
            let start = i;
            i += 1;
            while i < bytes.len() && (bytes[i] & 0xC0) == 0x80 && i - start < 4 {
                i += 1;
            }
            let piece = &bytes[start..i];
            if let Ok(s) = std::str::from_utf8(piece) {
                if let Some(id) = self.lookup(s) {
                    tokens.push(id);
                    continue;
                }
            }
            for &b in piece {
                tokens.push(i32::from(b) + 3);
            }
        }

        loop {
            let mut best_score = -1e10f32;
            let mut best_id = -1;
            let mut best_idx = None;
            for i in 0..tokens.len().saturating_sub(1) {
                let merged = format!(
                    "{}{}",
                    self.vocab[tokens[i] as usize],
                    self.vocab[tokens[i + 1] as usize]
                );
                if let Some(id) = self.lookup(&merged) {
                    let score = self.vocab_scores[id as usize];
                    if score > best_score {
                        best_score = score;
                        best_id = id;
                        best_idx = Some(i);
                    }
                }
            }
            let Some(idx) = best_idx else { break };
            tokens[idx] = best_id;
            tokens.remove(idx + 1);
        }

        if eos {
            tokens.push(EOS_ID);
        }
        tokens
    }

    /// llama3.cuda prompt compatibility patch for `"I have a dream"`.
    pub fn apply_dream_prompt_patch(tokens: &mut [i32]) {
        if tokens.len() > 1 && tokens[1] == 306 {
            tokens[1] = 76;
        }
    }

    pub fn decode(&self, prev_token: i32, token: i32) -> String {
        let mut piece = self.vocab.get(token as usize).cloned().unwrap_or_default();
        if prev_token == BOS_ID && piece.starts_with(' ') {
            piece.remove(0);
        }
        if let Some(byte) = parse_byte_token(&piece) {
            return (byte as char).to_string();
        }
        piece
    }

    /// Match `safe_printf` in llama3.cuda: skip empty / non-printable single bytes,
    /// plus the CJK byte-pair rewrite.
    pub fn printable_piece(piece: &str) -> String {
        if piece.is_empty() {
            return String::new();
        }
        let bytes = piece.as_bytes();
        if bytes.len() == 1 {
            let b = bytes[0];
            if !(b.is_ascii_graphic() || b.is_ascii_whitespace()) {
                return String::new();
            }
        }
        if bytes.len() >= 2 {
            match bytes[0] {
                0xC3 => return ((bytes[1] | 0x40) as char).to_string(),
                0xC2 => return (bytes[1] as char).to_string(),
                _ => {}
            }
        }
        piece.to_string()
    }
}

fn parse_byte_token(piece: &str) -> Option<u8> {
    let rest = piece.strip_prefix("<0x")?.strip_suffix('>')?;
    u8::from_str_radix(rest, 16).ok()
}

pub fn sample_argmax(logits: &[f32]) -> i32 {
    let mut max_i = 0;
    let mut max_p = logits[0];
    for (i, &p) in logits.iter().enumerate().skip(1) {
        if p > max_p {
            max_i = i;
            max_p = p;
        }
    }
    max_i as i32
}

#[cfg(test)]
mod tests {
    use super::*;

    fn toy_tokenizer() -> Tokenizer {
        // ids: 0=<unk> 1=<s> 2=</s> then pieces used by encode/merge
        let vocab = vec![
            "<unk>".into(),
            "<s>".into(),
            "</s>".into(),
            "a".into(),
            "b".into(),
            "ab".into(),
            " ".into(),
        ];
        let scores = vec![0.0, 0.0, 0.0, 1.0, 1.0, 10.0, 0.0];
        Tokenizer::from_parts(vocab, scores, 8)
    }

    #[test]
    fn encodes_with_bos_dummy_prefix_and_merge() {
        let tok = toy_tokenizer();
        let ids = tok.encode("ab", true, false);
        assert_eq!(ids[0], 1);
        assert_eq!(ids[1], 6, "dummy prefix is the space token");
        assert_eq!(ids[2], 5, "a+b should merge to ab");
    }

    #[test]
    fn decode_strips_space_after_bos() {
        let tok = toy_tokenizer();
        assert_eq!(tok.decode(1, 6), "");
        assert_eq!(tok.decode(0, 6), " ");
    }

    #[test]
    fn dream_patch_rewrites_token_one() {
        let mut ids = vec![1, 306, 40];
        Tokenizer::apply_dream_prompt_patch(&mut ids);
        assert_eq!(ids[1], 76);
        let mut other = vec![1, 77];
        Tokenizer::apply_dream_prompt_patch(&mut other);
        assert_eq!(other[1], 77);
    }

    #[test]
    fn argmax_picks_first_on_ties() {
        assert_eq!(sample_argmax(&[1.0, 3.0, 3.0, 2.0]), 1);
    }

    #[test]
    fn byte_token_decode() {
        let vocab = vec!["<unk>".into(), "<s>".into(), "</s>".into(), "<0x41>".into()];
        let tok = Tokenizer::from_parts(vocab, vec![0.0; 4], 8);
        assert_eq!(tok.decode(0, 3), "A");
    }
}

File: tools/bench/pytorch/tokenizer.py
"""llama2.c BPE tokenizer plus llama3.cuda host compatibility helpers."""

from __future__ import annotations

import pathlib
import struct
from dataclasses import dataclass

BOS_ID = 1
EOS_ID = 2


def apply_dream_prompt_patch(tokens: list[int]) -> None:
    if len(tokens) > 1 and tokens[1] == 306:
        tokens[1] = 76


def printable_piece(piece: str) -> str:
    if not piece:
        return ""
    raw = piece.encode("utf-8", errors="surrogateescape")
    if len(raw) == 1:
        b = raw[0]
        if not (32 <= b < 127 or b in (9, 10, 13, 32) or chr(b).isspace()):
            return ""
    if len(raw) >= 2:
        if raw[0] == 0xC3:
            return chr(raw[1] | 0x40)
        if raw[0] == 0xC2:
            return chr(raw[1])
    return piece


def _parse_byte_token(piece: str) -> int | None:
    if piece.startswith("<0x") and piece.endswith(">") and len(piece) == 6:
        try:
            return int(piece[3:5], 16)
        except ValueError:
            return None
    return None


@dataclass
class Tokenizer:
    vocab: list[str]
    vocab_scores: list[float]
    max_token_length: int
    sorted: list[tuple[str, int]]

    @classmethod
    def from_path(cls, path: str | pathlib.Path, vocab_size: int) -> Tokenizer:
        path = pathlib.Path(path)
        with path.open("rb") as f:
            max_token_length = struct.unpack("<I", f.read(4))[0]
            vocab: list[str] = []
            scores: list[float] = []
            for i in range(vocab_size):
                score = struct.unpack("<f", f.read(4))[0]
                (length,) = struct.unpack("<i", f.read(4))
                if length < 0:
                    raise ValueError(f"tokenizer piece {i} has negative length {length}")
                buf = f.read(length)
                vocab.append(buf.decode("utf-8", errors="surrogateescape"))
                scores.append(score)
        return cls.from_parts(vocab, scores, max_token_length)

    @classmethod
    def from_parts(
        cls, vocab: list[str], vocab_scores: list[float], max_token_length: int
    ) -> Tokenizer:
        sorted_vocab = sorted((s, i) for i, s in enumerate(vocab))
        return cls(vocab, vocab_scores, max_token_length, sorted_vocab)

    def lookup(self, s: str) -> int | None:
        lo, hi = 0, len(self.sorted)
        while lo < hi:
            mid = (lo + hi) // 2
            if self.sorted[mid][0] < s:
                lo = mid + 1
            else:
                hi = mid
        if lo < len(self.sorted) and self.sorted[lo][0] == s:
            return self.sorted[lo][1]
        return None

    def encode(self, text: str, bos: bool = True, eos: bool = False) -> list[int]:
        tokens: list[int] = []
        if bos:
            tokens.append(BOS_ID)
        if text:
            dummy = self.lookup(" ")
            if dummy is not None:
                tokens.append(dummy)
        raw = text.encode("utf-8")
        i = 0
        while i < len(raw):
            start = i
            i += 1
            while i < len(raw) and (raw[i] & 0xC0) == 0x80 and i - start < 4:
                i += 1
            piece = raw[start:i]
            try:
                s = piece.decode("utf-8")
            except UnicodeDecodeError:
                s = None
            if s is not None:
                found = self.lookup(s)
                if found is not None:
                    tokens.append(found)
                    continue
            tokens.extend(b + 3 for b in piece)

        while True:
            best_score = -1e10
            best_id = -1
            best_idx = None
            for i in range(len(tokens) - 1):
                merged = self.vocab[tokens[i]] + self.vocab[tokens[i + 1]]
                found = self.lookup(merged)
                if found is not None and self.vocab_scores[found] > best_score:
                    best_score = self.vocab_scores[found]
                    best_id = found
                    best_idx = i
            if best_idx is None:
                break
            tokens[best_idx] = best_id
            del tokens[best_idx + 1]

        if eos:
            tokens.append(EOS_ID)
        return tokens

    def decode(self, prev_token: int, token: int) -> str:
        piece = self.vocab[token] if 0 <= token < len(self.vocab) else ""
        if prev_token == BOS_ID and piece.startswith(" "):
            piece = piece[1:]
        byte_val = _parse_byte_token(piece)
        if byte_val is not None:
            return chr(byte_val)
        return piece


def sample_argmax(logits) -> int:
    """Greedy argmax over host logits; the lowest index wins ties (llama3.cuda `>` scan)."""
    if hasattr(logits, "argmax") and getattr(logits, "ndim", None) == 1:
        # torch/numpy argmax both return the first maximal index.
        return int(logits.argmax())
    best_i = 0
    best = float(logits[0])
    for i, value in enumerate(logits):
        v = float(value)
        if v > best:
            best = v
            best_i = i
    return best_i
